<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: Python Coding Exercises

_Authors: Joseph Nelson (DC) _

---

The following code challenges are drawn from common exercises used in technical interviews.

Please note that there may be several ways to approach each challenge. If you get stuck, try mapping out your approach in pseudocode first. Finally, while solutions to problems like these may be found online, remember that if you copy/paste code that you can't explain, you'll be missing out on the point of the project. The only way to truly learn a new skill is through practice, trial, and error - we can only help you improve by understanding where you are having trouble.

### Challenge 1: Largest Palindrome
A palindromic number reads the same both ways. For example, 1234321 is a palindrome. The largest palindrome made from the product of two two-digit numbers is 9009 = 91 × 99. Find the largest palindrome made from the product of two three-digit numbers. Afterward, write a brief explanation walking through your code's logic in markdown.

In [11]:
# function 1
def reverse_string(string):
    string_length = len(string)
    string_reversed=""
    for index in range(string_length-1,-1,-1):
        string_reversed += string[index]
    return string_reversed

# function 2
def is_palindrome(string):
    return string == reverse_string(string)


#code snippet to calculate largest panindrome
largest_palindrome = [[0,0,0]]

for number_1 in range(100,1000):
    for number_2 in range(100,1000):
            value = number_1 * number_2
            if(is_palindrome(str(value)) and value > largest_palindrome[0][0] ):
               largest_palindrome[0] = list((value,number_1,number_2))


print('Largest palindrome made from the product of 2 three-digit numbers is',largest_palindrome[0])




 largest palindrome made from the product of 2 three-digit numbers is [906609, 913, 993]


Answer is 906609. This value is a result of 913 * 993.

I have writtten 2 supporting functions and a code snippetto identify the answer. 

function 1 - reverse_string
   - this functions takes a string value as input parameter and retuns the reverse value of that string. 
   - it implements a for loop to navigate through the input parameter character by character in reverse order, using the range function with a step alue of -1.
   
function 2 - is_palindrome
   - this function takes a string value as input parameter and retuns boolean value indicating if the input parameter is a palindrome or not.
   - it calls the function 'reverse_string' to get the reverse value of the input parameter and compares the two to check if they are identical
   
code snippet
   - this code snippenis used to calculate and identify the largest palindrome made from the product of two three-digit numbers.
   - the smallest 3 digit number is 100 and the largest is 999
   - 2 nested for loops are used to calculate the product of every possible combination of 2 three-digit numbers. 
   - each for loop use the range function with a default step value of 1 to loop throug all the possible conbinations between 100 and 999, inclusive.
   - product of every combination is checked to see if that value is a panindrome by calling the 'is_palindrome'function and if so this value id greater than the palindrome identified so far 
   - if the product of the current combinaton is a palindrome and is greater than the previously held value, this value and the 2 three-digit numbers used to calculate the value are stored in a list object 'largest_palindrome'
   - length of the list object is 1 and it implements the concept of 'list of list' to hold the product and the 2  three-digit numbers used to caculate the product
   - the values stored in the list object 'largest_palindrome' when the 2 nested for loops completes execution is considered the largest palindrome made from the product of two three-digit numbers.





### Challenge 2: Summation of Primes
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17. Find the sum of all the primes below 2,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [10]:
def is_prime(number):
    if number >= 2:
        for value in range(2,number):
            if (number%value)==0:
                return False
    else:
        return False
    return True

sum_of_primes = 0
    
for number in range(2,2000):
    if is_prime(number):
        sum_of_primes +=number

print('Sum of primes below 2000 is',sum_of_primes)

sum of primes below 2000 is 277050


Answer is 277050.

I have writtten a supporting functions and a code snippetto compute the answer. 

supporting function - is_prime
   - this functions takes a number as an input parameter and returns true is its a prime number and fale if not. 
   - it first check if the input parameter is number 2 or a higher value, which is the basic condition for a prime number
   - then it runs a for loop for the range between 2 & the input parameter, calculating the modulo to fid out if any number between 2 and the input prameter will devide the number evenly without remainder. i.e. if any combination found to match this the input prameter is not a pirme number
    
    
code snippet
   - first it declares a variable to hold the sum of the prime numbers and initialized it to 0
   - then it runs a for loop between 2 and 2000 calling the is_prime function to check if each number in the for loop iteration step is a prime number or not. 
   - if a prime number is found its added to the variabe 'sum_of_primes'
   - finally, upon exit of the for loop the 'sum_of_primes' is displayed
    
    



### Challenge 3: Multiples of 3 and 5
If we list all of the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6, and 9. The sum of these multiples is 23. Find the sum of all the multiples of 3 or 5 below 1,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [6]:
sum_of_multiples = 0
for number in range(0,1000):
    if (number%3 == 0 or number%5 == 0):
        sum_of_multiples +=number

print('Sum of all the multiples of 3 or 5 below 1000 is',sum_of_multiples)


Sum of all the multiples of 3 or 5 below 1000 is 233168


Answer is 233168.

I have writtten a simple a code snippet to compute the answer.

   - first it declares a variable 'sum_of_multiples' to hold the sum of the matching natural numbers and initialized it to 0
   - then it runs a for loop between the range 0 and 1000.
   - each number in this range is checked to see if it could be devided by 3 or 5 without any remainders using the modulo (%) operator. If so, the number is added to the variable 'sum_of_multiples'
   - finally, upon exit of the for loop the 'sum_of_multiples' is displayed


### Challenge 4: String Compressor
Implement a method to perform basic string compression using the counts of repeated characters. (This is called run-length encoding.) For example, the string "aabcccccaaa" would become a2b1c5a3. If the “compressed” string would not become smaller than the original string, your method should return the original string. You can assume the string has only uppercase and lowercase letters (a–z). Specify whether your solution is case sensitive or case insensitive and what you would need to change to make it the other. Afterward, write a brief explanation walking through your code's logic in markdown.

In [42]:


def compress_string(string_to_compress,is_case_sensitive):

    letters = 'abcdefghijklmnopqrstuvwxyz'
    string_compressed = ''
    
    for char in string_to_compress:
        if(char.lower() not in letters):
            raise ValueError
 
    if len(string_to_compress) <= 1:
        return string_compressed

    if(not is_case_sensitive):
        string_to_compress = string_to_compress.lower()
    
    previous_char = string_to_compress[0]
    char_count = 1

    for pos in range(1,len(string_to_compress)):
        if(string_to_compress[pos]==previous_char):
            char_count += 1
        else:
            if(char_count ==1):
                string_compressed += string_to_compress[pos-1]
            else:
                string_compressed += string_to_compress[pos-1]+ str(char_count)
                char_count = 1
            previous_char = string_to_compress[pos]

    if(char_count ==1):
        string_compressed += string_to_compress[pos]
    else:
        string_compressed += string_to_compress[pos-1]+ str(char_count)


    if(len(string_to_compress)<= len(string_compressed)):
        return string_to_compress
    else:
        return string_compressed
            

print(compress_string('aasddDDfghhhhjkwwweev',False))




a2sd4fgh4jkw3e2v


I have written a function 'compress_string' to implement this string compression
 - this function takes 2 input parameters, first is the str type parameter 'string_to_compress' representing the string to compress and the second 'is_case_sensitive' a boolean type parameter indicting if the compression is case sensitive or not
 - within the function 
       - first all valid alphabet letters are assigned to a variable 'letter' to be used for valdation purposes
       - then a new varable to hold the compressed string 'string_compressed' is declared and initialised to empty.
       - then, the code checks if the input parameter 'string_to_compress' includes any not aphabet charaters and if so exit the function by raising a ValueError
       - then, the code checks if the legnth of the string_to_compress' is lessthan or equal to 1, if so there is no need to compress the string. Thus returns the original string and exit the function
       - then, if the 'is_case_sensitive' is set to false the 'string_to_compress' id converted to lower case
       - then variables 'previous_char' and 'char_count' are initialised with the first character of the 'string_to_compress' and character count to 1 respectively to be used within the following for loop. 
       - then a for loop id implemented to raverse through the 'string_to_compress' from its 2nd character (as first is already loaded above) to the last 
       - within the for loop, 
             - if the previous character is same as current character the character counter is increased by 1. 
             - else, if the character count of the current character is 1 ( i.e. current character has not repeated), current character is apended to the variable 'string_compressed'
             - else (i.e character has repeasted), the current character in concerned and the count of the occurences of that character is apended to the variable 'string_compressed'. 'char_count' is also reset to 1 to start counting of a new character series.
             - previouse_char is then assinged with the current char inpreperation for the next iteration for the for loop
       - same condition above is implemented outside the for loop to ensure last character set to compressed correctly. 
       - finally, the lenght of the variables 'string_to_compress' and 'string_compressed' are compared and if the lenght of the original is less than or same as the compressed, original string is returned. else the compressed string is returned.
             
 

### Challenge 5: FizzBuzz

Write a program that prints all of the numbers from 1 to 100. For multiples of 3, instead of the number, print "Fizz;" for multiples of 5, print "Buzz." For numbers that are multiples of both 3 and 5, print "FizzBuzz." Afterward, write a brief explanation walking through your code's logic in markdown.

In [1]:

for number in range(1,101):
    if(number%3 == 0 and number%5==0):
        print('FizzBuzz')
    elif(number%3 == 0):
        print('Fizz')
    elif(number%5 == 0):
        print('Buzz')
    else:
        print(number)

1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz
16
17
Fizz
19
Buzz
Fizz
22
23
Fizz
Buzz
26
Fizz
28
29
FizzBuzz
31
32
Fizz
34
Buzz
Fizz
37
38
Fizz
Buzz
41
Fizz
43
44
FizzBuzz
46
47
Fizz
49
Buzz
Fizz
52
53
Fizz
Buzz
56
Fizz
58
59
FizzBuzz
61
62
Fizz
64
Buzz
Fizz
67
68
Fizz
Buzz
71
Fizz
73
74
FizzBuzz
76
77
Fizz
79
Buzz
Fizz
82
83
Fizz
Buzz
86
Fizz
88
89
FizzBuzz
91
92
Fizz
94
Buzz
Fizz
97
98
Fizz
Buzz


I have writtten a simple a code snippet to display(print) the required output
  -this code runs a for loop for the range from 1 to 101
  -each number in the range is 
      - first tested to see if its a multiplication of both 3 & 5 and if so, 'FizzBuzz' is displayed(printed) instead of the number
      - if this first condition is not met (i.e. first if condition) and if the number is a multiplication of 3, 'Fizz' is displayed(printed) instead of the number
      - if the second condition is also not met (i.e. first elif condition) and if the number is a multiplication of 5, 'Buzz' is displayed(printed) instead of the number
      - if the third condition is also not met (i.e. second elif condition) the number is not a multiplication of 3 or 5. Hence the number is displayed(printed)
  
  